File: packages/react-native-fast-io/android/src/main/java/com/margelo/nitro/fastio/HybridDuplexStream.kt
package com.margelo.nitro.fastio

import java.io.PipedInputStream
import java.io.PipedOutputStream

class HybridDuplexStream : HybridDuplexStreamSpec() {
    private val pipeIn = PipedInputStream(HybridStreamFactory.BUFFER_SIZE)
    private val pipeOut = PipedOutputStream(pipeIn)

    override var inputStream: HybridInputStreamSpec = HybridInputStream(pipeIn).also {
        System.err.println("Created input stream wrapper")
    }
    
    override var outputStream: HybridOutputStreamSpec = HybridOutputStream(pipeOut).also {
        System.err.println("Created output stream wrapper")
    }

    override val memorySize: Long = 0L

    fun close() {
        try {
            System.err.println("Closing duplex stream")
            pipeOut.close()
            pipeIn.close()
        } catch (e: Exception) {
            System.err.println("Error closing duplex stream: ${e.message}")
        }
    }
} 

File: packages/react-native-fast-io/ios/HybridDuplexStream.swift
//
//  HybridStreamManager.swift
//  FastIO
//
//  Created by Mike Grabowski on 07/11/2024.
//

import Foundation

class HybridDuplexStream : HybridDuplexStreamSpec {
  var inputStream: (any HybridInputStreamSpec)
  var outputStream: (any HybridOutputStreamSpec)
  
  override init() {
    var inputStreamRef: InputStream? = InputStream()
    var outputStreamRef: OutputStream? = OutputStream(toMemory: ())
    
    Stream.getBoundStreams(withBufferSize: Int(HybridStreamFactory.BUFFER_SIZE), inputStream: &inputStreamRef, outputStream: &outputStreamRef)
    
    guard let inputStreamRef, let outputStreamRef else {
      fatalError("Could not create streams")
    }
    
    inputStream = HybridInputStream(stream: inputStreamRef)
    outputStream = HybridOutputStream(stream: outputStreamRef)
  }
}
